packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -554,212 +584,210 @@ function basicStateReducer<S>(state: S, action: BasicStateAction<S>): S {
+  initialAction: void | null | A,
 ): [S, Dispatch<A>] {
-  if (__DEV__) {
-    if (reducer !== basicStateReducer) {
-      currentHookNameInDev = 'useReducer';
-    }
-  }
-  currentlyRenderingFiber = resolveCurrentlyRenderingFiber();
-  workInProgressHook = createWorkInProgressHook();
-  if (__DEV__) {
-    currentHookNameInDev = null;
+  const hook = mountWorkInProgressHook();
+  // TODO: Lazy init API will change before release.
+  if (initialAction !== undefined && initialAction !== null) {
+    // $FlowFixMe - Must express with overloading.
+    initialState = reducer(initialState, initialAction);
   }
-  let queue: UpdateQueue<S, A> | null = (workInProgressHook.queue: any);
-  if (queue !== null) {
-    // Already have a queue, so this is an update.
-    if (numberOfReRenders > 0) {
-      // This is a re-render. Apply the new render phase updates to the previous
-      // work-in-progress hook.
-      const dispatch: Dispatch<A> = (queue.dispatch: any);
-      if (renderPhaseUpdates !== null) {
-        // Render phase updates are stored in a map of queue -> linked list
-        const firstRenderPhaseUpdate = renderPhaseUpdates.get(queue);
-        if (firstRenderPhaseUpdate !== undefined) {
-          renderPhaseUpdates.delete(queue);
-          let newState = workInProgressHook.memoizedState;
-          let update = firstRenderPhaseUpdate;
-          do {
-            // Process this render phase update. We don't have to check the
-            // priority because it will always be the same as the current
-            // render's.
-            const action = update.action;
-            if (__DEV__) {
-              enterDisallowedContextReadInDEV();
-              isInHookUserCodeInDev = true;
-            }
-            newState = reducer(newState, action);
-            if (__DEV__) {
-              exitDisallowedContextReadInDEV();
-              isInHookUserCodeInDev = false;
-            }
-            update = update.next;
-          } while (update !== null);
-
-          workInProgressHook.memoizedState = newState;
-
-          // Don't persist the state accumlated from the render phase updates to
-          // the base state unless the queue is empty.
-          // TODO: Not sure if this is the desired semantics, but it's what we
-          // do for gDSFP. I can't remember why.
-          if (workInProgressHook.baseUpdate === queue.last) {
-            workInProgressHook.baseState = newState;
-          }
+  hook.memoizedState = hook.baseState = initialState;
+  const queue = (hook.queue = {
+    last: null,
+    dispatch: null,
+    eagerReducer: reducer,
+    eagerState: (initialState: any),
+  });
+  const dispatch: Dispatch<A> = (queue.dispatch = (dispatchAction.bind(
+    null,
+    // Flow doesn't know this is non-null, but we do.
+    ((currentlyRenderingFiber: any): Fiber),
+    queue,
+  ): any));
+  return [hook.memoizedState, dispatch];
+}
+
+function updateReducer<S, A>(reducer: (S, A) => S): [S, Dispatch<A>] {
```

Comment:
This function gets exposed to always be called with three args. I'm not sure how much it matters in practice but just to be safe, we should try to ensure that the arg count is consistent so let's add the unused args here and pass them in to ensure type consistency. Same thing with useState...